### PR TITLE
Add `test_smoke` test re-open

### PR DESCRIPTION
**Pull request by BON4**
**Issue https://github.com/canonical/postgresql-k8s-operator/issues/468**

This pull request is based on the [same PR](https://github.com/canonical/postgresql-operator/pull/454) but for VM postgresql-operator

All this cases should be covered:

- Test Application. VM resources is removed after application removal. Install app, remove app (removal test), check no garbage left.
- Test Unit. Re-Installation. Remove unit with `--force --no-wait`, storage should be detached.
- Test Unit. Garbage ignorance. Charm should deploy in dirty environment with garbage storage.
- Test Application. Application should deploy in dirty environment with garbage storage from another application.

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]
@BON4 Thank you! I would like to merge this, can you please rebase PR to resolve conflicts. Thank you!

CC: @marceloneppel can you please review this to merge and sync VM and K8s tests. Tnx!

File: tests/integration/ha_tests/helpers.py [hunk 1]
```diff
@@ -759,3 +773,229 @@ async def stop_continuous_writes(ops_test: OpsTest) -> int:
...
+            f"Pod: {pod.metadata.name} STATUS: {pod.status.phase} TAGGED: {unit_name.replace('/', '-')}"
+        )
+        if (pod.metadata.name == unit_name.replace("/", "-")) and (pod.status.phase == "Running"):
+            return True
+
+    return False
+
+
+async def is_storage_exists(
+    ops_test: OpsTest, storage_id: str, include_detached: bool = False
```
Comment (marceloneppel) [inline tests/integration/ha_tests/helpers.py:816]: The `include_detached` parameter is not being used in this function.

File: tests/integration/ha_tests/test_smoke.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+)
+
+TEST_DATABASE_RELATION_NAME = "test_database"
+DUP_DATABASE_APP_NAME = DATABASE_APP_NAME + "2"
+
+logger = logging.getLogger(__name__)
+
+env = os.environ
+env["KUBECONFIG"] = os.path.expanduser("~/.kube/config")
+print(f"Model Name: {DATABASE_APP_NAME}")
```
Comment (marceloneppel) [inline tests/integration/ha_tests/test_smoke.py:43]: This `print` statement may be removed. I think it was used only for testing.

File: tests/integration/ha_tests/helpers.py [hunk 3]
```diff
@@ -431,7 +439,13 @@ async def is_connection_possible(ops_test: OpsTest, unit_name: str) -> bool:
...
                 connection.close()
                 return success
-    except (psycopg2.Error, RetryError):
+    except (psycopg2.Error, RetryError) as e:
+        if isinstance(e, RetryError):
+            last_attempt = e.last_attempt
+            last_exception = last_attempt.exception()
+            print(f"is_connection_possible err: {e}, last exception: {last_exception}")
+        else:
+            print(f"is_connection_possible err: {e}")
```
Comment (marceloneppel) [inline tests/integration/ha_tests/helpers.py:448]: Let's replace all these calls to `print` with `logger.error` to follow the pattern we already have.

**Review by marceloneppel — CHANGES_REQUESTED** [review]
Thanks for the contribution, @BON4! I requested some minor changes in the code.

File: tests/integration/ha_tests/test_smoke.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+        # Force remove unit without storage removal
+        logger.info("scale to 0 with force")
+        await remove_unit_force(ops_test, 1)
+
+        # Storage will remain with deatached status
+        logger.info("werifing is storage exists")
+        for attempt in Retrying(stop=stop_after_delay(15 * 3), wait=wait_fixed(3), reraise=True):
+            with attempt:
+                assert await is_storage_exists(ops_test, storage_id, include_detached=True)
```
Comment (marceloneppel) [inline tests/integration/ha_tests/test_smoke.py:102]: `include_detached` was removed from the helper function, but it's still here.

**Review by marceloneppel — CHANGES_REQUESTED** [review]
One last change. Apart from that, LGTM, as soon as tests pass.

File: tests/integration/ha_tests/helpers.py (hunk 1 above)
Comment (BON4) [reply tests/integration/ha_tests/helpers.py]: fixed

File: tests/integration/ha_tests/test_smoke.py (hunk 2 above)
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: fixed

File: tests/integration/ha_tests/helpers.py (hunk 3 above)
Comment (BON4) [reply tests/integration/ha_tests/helpers.py]: fixed

File: tests/integration/ha_tests/test_smoke.py (hunk 4 above)
Comment (BON4) [reply tests/integration/ha_tests/test_smoke.py]: Removed it

**Review by marceloneppel — APPROVED** [review]

**Comment by BON4** [comment]
@marceloneppel could you re-ran checks? CLA check, somehow, pulled my email from github account (not from commits). I made my email public, it should pass.

**Review by marceloneppel — APPROVED** [review]
LGTM! As the CLA check was the only test that was executed in the last commit, I tested this PR manually, as well as the failing self-healing tests, to ensure we're not breaking something.

**Pull request merged**